### PR TITLE
Add PM Target Cycling to Chat History plugin

### LINKED ISSUE #7885: Friends chat bubbles
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. Ex. I'm always
frustrated when [...]
Im always frustrated when i press tab while talking to friends and i respond to the wrong person or it gets cut off.
**Describe the solution you'd like**
chat head bubbles replacing bubbles on bottom for "game, public, private, clan, trade" as the friend name and a chat filtering to only show their messages
**Describe alternatives you've considered**
A clear and concise description of any alternative solutions or features you've
considered.
not sure how tab will react when using this tool, may need an alternativate to work around it
**Additional context**

### PR DESCRIPTION
**Pull request by TheStonedTurtle**
Pressing tab again while typing a private message will change the target recipient by cycling through the list of recipients

Closes #7885

### CODE REVIEW

**Comment by TheStonedTurtle** [comment]
PR has been updated to invoke the jagex script when changing targets and to grab the current target name via the `VarClientStrChanged` event

File: runelite-api/src/main/java/net/runelite/api/ScriptID.java [hunk 1]
```diff
@@ -45,6 +45,16 @@
 	 */
 	public static final int CHATBOX_INPUT = 96;
 
+	/**
+	 * Opens the Private Message chat interface
+	 *
+	 * Jagex refers to this script as {@code meslayer_mode6}
+	 * <ul>
+	 * <li> String Player to send private message too</li>
```
Comment (Abextm) [inline runelite-api/src/main/java/net/runelite/api/ScriptID.java]:
```suggestion
	 * <li> String Player to send private message to </li>
```

File: runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		if (e.getKeyCode() == HOTKEY)
+		{
+			if (target == null || friends.size() == 0)
+			{
+				return;
+			}
+
+			if (client.getVar(VarClientInt.INPUT_TYPE) == InputType.PRIVATE_MESSAGE.getType())
+			{
+				String newFriend = friends.lower(target);
```
Comment (Abextm) [inline runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java]: Anything touching `friends` should happen on the clientThread

**Comment by TheStonedTurtle** [comment]
The requested changes have been made also using a LinkedHashSet instead of a TreeSet

**Comment by TheStonedTurtle** [comment]
I've changed this to loop over the LinkedHashSet instead of converting it to an ArrayList.

**Comment by TheStonedTurtle** [comment]
Rebased PR & Cleaned up commit history

File: runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+import net.runelite.client.input.KeyListener;
+import net.runelite.client.input.KeyManager;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.client.util.Text;
+
+@Slf4j
+@PluginDescriptor(
+	name = "Private Message Cycle",
+	description = "When sending a private message pressing the tab button again will cycle through your previous recipients"
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java]: This description should be shorter and make more sense

File: runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	@Inject
+	private Client client;
+
+	@Inject
+	private ClientThread clientThread;
+
+	@Inject
+	private KeyManager keyManager;
+
+	private final LinkedHashSet<String> friends = new LinkedHashSet<>();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java]: This should probably just be a fixed size `Deque`. It certainly should not grow unbounded like you have here.

File: runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+				friends.add(name);
+		}
+	}
+
+	@Subscribe
+	public void onVarClientStrChanged(VarClientStrChanged c)
+	{
+		if (c.getIndex() == VarClientStr.PRIVATE_MESSAGE_TARGET.getIndex())
+		{
+			target = client.getVar(VarClientStr.PRIVATE_MESSAGE_TARGET);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java]: Why can't this just be fetched in `keyPressed`?

**Comment by Kamielvf** [comment]
This should go into the `Chat History` plugin IMO.

**Comment by TheStonedTurtle** [comment]
Implemented Review changes and now conserves message content when changing targets

Also rebased to use `ChatMessage` since `SetMessage` was removed.

**Comment by ShaunDreclin** [comment]
Would this let you reply to the last pm sent to you even if you're not currently sending one? Being able to just hit tab to reply to people would be awesome

**Comment by TheStonedTurtle** [comment]
> Being able to just hit tab to reply to people would be awesome

This is a feature jagex already has implemented. Have you gone this whole time right-click responding to PMs?

**Comment by ShaunDreclin** [comment]
No of course not

🙄

**Comment by Adam-** [comment]
I agree this would make the most sense as part of the chathistory plugin.

**Comment by TheStonedTurtle** [comment]
I'll add this to the ChatHistory plugin then. I assume this functionality should have a config option?

**Comment by Hydrox6** [comment]
If it's going to be added there, can a toggle be added to not have the chat history itself?

File: runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	@Inject
+	private Client client;
+
+	@Inject
+	private ClientThread clientThread;
+
+	@Inject
+	private KeyManager keyManager;
+
+	private final EvictingQueue<String> friends = EvictingQueue.create(100);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java]: this should be ~5

**Comment by TheStonedTurtle** [comment]
Updated PR to add the PM Cycling functionality to the Chat History plugin

Gif after update:
https://gfycat.com/MildCompetentGourami

File: runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryConfig.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	)
+	default boolean retainChatHistory()
+	{
+		return true;
+	}
+
+	@ConfigItem(
+		keyName = "pmTargetCycling",
+		name = "PM Target Cycling",
+		description = "Retains chat history when logging in/out or world hopping",
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryConfig.java]: This needs an updated description

File: runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java [hunk 8]
```diff
@@ -57,22 +73,53 @@
 		ChatMessageType.GAME
 	);
 
+	private static final int CYCLE_HOTKEY = KeyEvent.VK_TAB;
+	private static final Set<ChatMessageType> CYCLE_TYPES = Sets.newHashSet(
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java]: `ImmutableSet`

File: runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java [hunk 9]
```diff
@@ -92,7 +139,7 @@ public void onChatMessage(ChatMessage chatMessage)
 			return;
 		}
 
-		if (ALLOWED_HISTORY.contains(chatMessage.getType()))
+		if (config.retainChatHistory() && ALLOWED_HISTORY.contains(chatMessage.getType()))
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java]: This config opt should be on the replay.

File: runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java [hunk 10]
```diff
@@ -143,4 +199,64 @@ private static String nbsp(final String message)
 
 		return null;
 	}
+
+	@Override
+	public void keyPressed(KeyEvent e)
+	{
+		if (!config.pmTargetCycling() || e.getKeyCode() != CYCLE_HOTKEY)
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java]: Check keycode first.

File: runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java [hunk 11]
```diff
@@ -143,4 +199,64 @@ private static String nbsp(final String message)
...
+
+		if (client.getVar(VarClientInt.INPUT_TYPE) != InputType.PRIVATE_MESSAGE.getType())
+		{
+			return;
+		}
+
+		clientThread.invoke(() ->
+		{
+			final String currentTarget = client.getVar(VarClientStr.PRIVATE_MESSAGE_TARGET);
+			if (currentTarget == null || friends.size() == 0)
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java]: This should be `isEmpty` - and move it to `findPreviousFriend`

File: runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java (hunk 8 above)
Comment (TheStonedTurtle) [reply runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java]: I'll update the one right above this as well then.

**Comment by TheStonedTurtle** [comment]
Requested review changes have been implemented.

File: runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java [hunk 12]
```diff
@@ -57,30 +72,61 @@
...
+		keyManager.unregisterKeyListener(this);
 	}
 
 	@Subscribe
 	public void onChatMessage(ChatMessage chatMessage)
 	{
 		// Start sending old messages right after the welcome message, as that is most reliable source
 		// of information that chat history was reset
-		if (chatMessage.getMessage().equals(WELCOME_MESSAGE))
+		if (config.retainChatHistory() && chatMessage.getMessage().equals(WELCOME_MESSAGE))
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java]: This is a behavior change and makes it start retaining the welcome message. Maybe don't do this.

File: runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java [hunk 13]
```diff
@@ -108,6 +154,15 @@ public void onChatMessage(ChatMessage chatMessage)
 				messageQueue.offer(queuedMessage);
 			}
 		}
+
+		if (CYCLE_TYPES.contains(chatMessage.getType()))
+		{
+			final String name = Text.removeTags(chatMessage.getName());
+			// Remove to ensure uniqueness & it's place in history
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java]: `its`

File: runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java [hunk 14]
```diff
@@ -108,6 +154,15 @@ public void onChatMessage(ChatMessage chatMessage)
...
+			friends.remove(name);
+			friends.add(name);
+
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java]: Remove this line.

**Comment by TheStonedTurtle** [comment]
Fixed commits to address review comments

File: runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java [hunk 15]
```diff
@@ -108,6 +159,14 @@ public void onChatMessage(ChatMessage chatMessage)
 				messageQueue.offer(queuedMessage);
 			}
 		}
+
+		if (CYCLE_TYPES.contains(chatMessage.getType()))
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java]: This was just a switch in the other PR which seems like it is probably cleaner?

Comment (TheStonedTurtle) [reply runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java]: switched to a switch statement

**Pull request merged**